cpp/include/cudf/scalar/scalar.hpp

```diff
@@ -402,18 +403,44 @@ class timestamp_scalar : public detail::fixed_width_scalar<T> {
    * @param[in] data The scalar's data in device memory
    * @param[in] is_valid Whether the value held by the scalar is valid
    */
-  timestamp_scalar(rmm::device_scalar<T>&& data,
-                   bool is_valid                       = true,
-                   cudaStream_t stream                 = 0,
-                   rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource())
+  chrono_scalar(rmm::device_scalar<T>&& data,
+                bool is_valid                       = true,
+                cudaStream_t stream                 = 0,
+                rmm::mr::device_memory_resource* mr = rmm::mr::get_default_resource())
     : detail::fixed_width_scalar<T>(std::forward<rmm::device_scalar<T>>(data), is_valid, stream, mr)
   {
   }
 
   /**
    * @brief Return the duration in number of ticks since the UNIX epoch.
    */
-  typename T::duration::rep ticks_since_epoch() { return this->value().time_since_epoch().count(); }
+  template <typename ChronoT = T>
+  typename std::enable_if_t<
+    std::is_same<typename cudf::is_timestamp_t<ChronoT>::type, std::true_type>::value,
+    typename ChronoT::rep>
+  ticks_since_epoch()
+  {
+    return this->value().time_since_epoch().count();
+  }
+
+  template <typename ChronoT = T>
+  typename std::enable_if_t<
+    std::is_same<typename cudf::is_duration_t<ChronoT>::type, std::true_type>::value,
+    typename ChronoT::rep>
+  ticks_since_epoch()
+  {
+    return this->value().count();
+  }
+};
+
+template <typename T>
+struct timestamp_scalar : chrono_scalar<T, cudf::is_timestamp_t<T>> {
```

Comment:
template conditions and param seem bit complicated. 
check changes from #5394 and pull.

Reply:
resolved with merge from #5359